Issue 45722: Editable grid support for column fill actions (v1)

cnathe opened the pull request:
#### Rationale
https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=45722

Users expect Excel-like abilities from the app editable grid. This PR adds the "fill down via cell range drag handle" type action for a grid column to be able to copy a cell value down a column or to be able to fill a column down with a numeric sequence.

The fill sequence rules are:
- If the initSelection is for a single cell, the fill operation will always be a copy of that value.
- If the initSelection includes a range of cells and all values are numeric, fill via a generated sequence where the step/diff is based on the first and last value in the initSelection.
- If the initSelection includes a range of cells and not all values are numeric, fill via a copy of all of the values in initSelection.

Note that this initial implementation is just for a single column in the editable grid and does not take into account the field type for the given column (i.e. just based on the values in the cells that are selected when the drag starts).

#### Related Pull Requests
* https://github.com/LabKey/labkey-ui-components/pull/908
* https://github.com/LabKey/inventory/pull/494
* https://github.com/LabKey/biologics/pull/1458
* https://github.com/LabKey/sampleManagement/pull/1119

#### Changes
* initially only support for a single column selection range
* initial support for copying a cell value, or set of selected values, down a column on cell handle drag
* initial support for numeric fill sequence for a column if initSelection has int or decimal values
* allow cell key handler for CTRL+D to fill down a single column

cnathe committed:
v2.200.1-fb-editableGridFillCol45722.0

cnathe committed:
Issue 45722: Editable grid support for column fill actions
  * initial support for copying a cell value, or set of selected values, down a column on cell handle drag

cnathe committed:
only show cell drag handle for single column selection

cnathe committed:
Merge branch 'develop' into fb_editableGridFillCol45722

cnathe committed:
Merge branch 'develop' into fb_editableGridFillCol45722

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
initial support for numeric fill sequence for a column if initSelection has int or decimal values

cnathe committed:
v2.200.2-fb-editableGridFillCol45722.0

cnathe committed:
allow cell key handler for CTRL+D to fill down a single column

cnathe committed:
v2.200.2-fb-editableGridFillCol45722.1

cnathe committed:
fix for cell drag handle with numeric (right aligned) cells

cnathe committed:
v2.200.2-fb-editableGridFillCol45722.2

cnathe commented on packages/components/src/internal/components/editable/Cell.tsx:
```diff
@@ -15,23 +15,24 @@
 
-import { LookupCell, LookupCellProps } from './LookupCell';
-import { DateInputCell, DateInputCellProps } from './DateInputCell';
+import {LookupCell, LookupCellProps} from './LookupCell';
+import {DateInputCell, DateInputCellProps} from './DateInputCell';
```
not sure why the spaces on these imports got updated...I'll lint before merging.

cnathe committed:
Merge branch 'develop' into fb_editableGridFillCol45722

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md
#	packages/components/src/internal/components/editable/Cell.tsx

cnathe committed:
v2.201.1-fb-editableGridFillCol45722.0

labkey-susanh commented on packages/components/src/internal/actions.ts:
```diff
@@ -943,6 +966,82 @@ async function prepareInsertRowDataFromBulkForm(
+    const initCellRawValues = initCellValues.map(cellValue => cellValue?.first()?.raw);
+    const initCellDisplayValues = initCellValues.map(cellValue => cellValue?.first()?.display);
+
+    // use the display values to determine sequence type to account for lookup cell values with numeric key/raw values
```
Lookups are interesting.  I wonder if anyone would ever expect to be able to fill a sequence from a lookup using the ordering of the lookup values.  Surely not, but what is the expected behavior for a lookup. Perhaps this functionality just shouldn't be available?

cnathe replied:
Yeah, using a numeric incrementing/decrementing sequence to fill al lookup by key seems wrong. Currently, the implementation will treat lookups as text and copy values down the column.

labkey-susanh commented on packages/components/src/internal/actions.ts:
```diff
@@ -943,6 +966,82 @@ async function prepareInsertRowDataFromBulkForm(
+}
+
+// https://stackoverflow.com/questions/10713878/decimal-subtraction-problems-in-javascript
+function decimalDifference(first, second, substract = true): number {
```
```suggestion
function decimalDifference(first, second, subtract = true): number {
```

cnathe replied:
done

labkey-nicka commented on packages/components/src/internal/components/editable/Cell.tsx:
```diff
@@ -248,14 +254,19 @@ export class Cell extends React.PureComponent<Props, State> {
     handleSelect = (event): void => {
         const { cellActions, colIdx, rowIdx, selected } = this.props;
         const { selectCell } = cellActions;
+        const isDragHandle = event.target?.className?.indexOf(CELL_SELECTION_HANDLE_CLASSNAME) > -1;
```
Move computation of `isDragHandle` inside the `else if`. This will make it so we don't process this every time `handleSelect` is fired.

cnathe replied:
done

labkey-nicka commented on packages/components/src/internal/components/editable/EditableGrid.tsx:
```diff
@@ -592,6 +600,13 @@ export class EditableGrid extends PureComponent<EditableGridProps, EditableGridS
         } = this.props;
+
+        // initial implementation of drag handle fill actions only support single column selection
+        const sortedSelectionCellKeys = !editorModel.hasMultipleColumnSelection()
```
Consider making `lastSelection()` a method on `EditorModel`.

```ts
lastSelection(colIdx: number, rowIdx: number): boolean {
    let cellKeys = [];

    // Initial implementation of drag handle fill actions only support single column selection
    if (!this.hasMultipleColumnSelection()) {
        if (this.hasMultipleSelection()) {
            cellKeys = this.getSortedSelectionKeys();
        } else {
            cellKeys = [this.getSelectionKey()];
        }
    }

    if (cellKeys.length === 0) {
        return false;
    }

    return cellKeys.indexOf(genCellKey(colIdx, rowIdx)) === cellKeys.length - 1;
}
```

and then it can be removed from being an input to `inputCellFactory` resulting in a line that looks like:

```tsx
lastSelection={editorModel ? editorModel.lastSelection(colIdx, rn) : false}
```

cnathe replied:
Sounds good. Done (thanks for the patch)

labkey-nicka commented on packages/components/src/internal/events.ts:
```diff
@@ -46,6 +46,10 @@ export function isSelectAll(event: React.KeyboardEvent<any>): boolean {
     return isEvent(event) && event.keyCode === KEYS.A && (event.ctrlKey || event.metaKey);
 }
 
+export function isFillDown(event: React.KeyboardEvent<any>): boolean {
```
nit: Consider refactoring these events to:

```ts
function isEvent(event: React.SyntheticEvent<any>): boolean {
    return event !== undefined && event !== null;
}

function isMetaKeyEvent(event: React.KeyboardEvent<any>, keyCode: number): boolean {
    return isEvent(event) && event.keyCode === keyCode && (event.ctrlKey || event.metaKey);
}

export const isCopy = (event: React.KeyboardEvent<any>): boolean => isMetaKeyEvent(event, KEYS.C);
export const isFillDown = (event: React.KeyboardEvent<any>): boolean => isMetaKeyEvent(event, KEYS.D);
export const isPaste = (event: React.KeyboardEvent<any>): boolean => isMetaKeyEvent(event, KEYS.V);
export const isSelectAll = (event: React.KeyboardEvent<any>): boolean => isMetaKeyEvent(event, KEYS.A);
```

cnathe replied:
I like it. Thanks for the patch.

labkey-nicka commented on packages/components/src/internal/models.ts:
```diff
@@ -601,10 +604,22 @@ export class EditorModel
         return this.selectedColIdx > -1 && this.selectedRowIdx > -1;
     }
 
+    getSelectionKey(): string {
```
nit: consider `get selectionKey(): string`

cnathe replied:
done

labkey-nicka commented on packages/components/src/internal/models.ts:
```diff
@@ -613,6 +628,10 @@ export class EditorModel
         return colIdx > -1 && rowIdx > -1 && this.selectionCells.get(genCellKey(colIdx, rowIdx)) !== undefined;
     }
 
+    getSortedSelectionKeys(): string[] {
```
nit: consider `get sortedSelectionKey(): string[]`

cnathe replied:
done

cnathe committed:
Merge branch 'develop' into fb_editableGridFillCol45722

cnathe committed:
CR feedback
- editorModel.getSelectionKey() -> editorModel.selectionKey
- editorModel.getSortedSelectionKeys() -> editorModel.sortedSelectionKeys
- move isDragHandle calc into elseif
- move code to an editorModel.lastSelection function
- refactor event functions in events.ts

cnathe committed:
v2.201.1-fb-editableGridFillCol45722.1

cnathe committed:
new jest tests

cnathe committed:
v2.201.1-fb-editableGridFillCol45722.2

cnathe committed:
fix for cell to allow for hover display of error message with drag handle

cnathe committed:
v2.201.1-fb-editableGridFillCol45722.3

cnathe committed:
handle case where fill column action spans a readonly row

cnathe committed:
v2.201.1-fb-editableGridFillCol45722.4

cnathe committed:
handle case where fill column action spans a readonly cell

cnathe committed:
v2.201.1-fb-editableGridFillCol45722.5

cnathe committed:
no drag handle for readonly cell

cnathe committed:
v2.201.1-fb-editableGridFillCol45722.6

cnathe committed:
yarn run lint-branch-fix

cnathe committed:
Merge branch 'develop' into fb_editableGridFillCol45722

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
Update releaseNotes with version number and release date

cnathe committed:
v2.202.0
